### ISSUE TITLE
[Enchancement] Add more customization

### ISSUE THREAD

**Issue by SkullGamer205** [jekyllgrim/UniversalFlexibleHUD#13]
## **Add some settings for HUD elements:**
- [x] Add resizing of each element (#9) (Exist in latest dev build)
- [ ] Increase the shift limit of the horizontal and vertical shift of an element (Up to 4096) (Or add a shift by percentage)
- [ ] Add the ability to change the font of each element (not only from those existing in Doom, but from those loaded by separate WADs (By entering the font name))
- [ ] Ability to attach a character's face to the left of the character status information
- [ ] Ability to attach the character’s face in the middle (a.k.a. classic doom hud bar) (On the left is health, in the center is the face, on the right is armor)
![изображение](https://github.com/jekyllgrim/UniversalFlexibleHUD/assets/61167080/b9d447cd-5313-4808-9c07-171b3a0a0810) Like this
- [ ] Ability to change the size of health/armor icons
- [ ] Ability to display health/armor and ammo not only horizontally, but also vertically
![изображение](https://github.com/jekyllgrim/UniversalFlexibleHUD/assets/61167080/4f01f324-1d83-49aa-9dd6-4328ef96830a) Like this
- [ ] Possibility of adding several columns for all ammo display (In case there are many types of ammo, or for the compactness of the HUD)
![изображение](https://github.com/jekyllgrim/UniversalFlexibleHUD/assets/61167080/0bafda53-08c4-4cb9-a0df-288dadd6a57b) Like this (if 2 collums)
- [ ] Add the ability to display only current ammo, or current ammo with maximum ammo. (Screen above)
- [ ] Option to enable vertical inventory bar
- [ ] Ability to load minimap colors from the fullscreen map

**Comment by jekyllgrim** [comment]
> Increase the shift limit of the horizontal and vertical shift of an element (Up to 4096) (Or add a shift by percentage)

I'm not sure I see what the point of this is. The distance is designed to cover a specific region around the currently used anchor point. If you want to move further, it would make more sense to just change the anchor point first (move to a different corner/edge).

> Add the ability to change the font of each element (not only from those existing in Doom, but from those loaded by separate WADs (By entering the font name))

Currently being worked on. Not for each element separately, though, that'd be too much work. There are 3 fonts used in the HUD, and each of them will be modifiable separately, with support for custom font names. Even that was kind of a nightmare to implement, so I'm not sure I will expand this further.

> Ability to attach a character's face to the left of the character status information
> Ability to attach the character’s face in the middle (a.k.a. classic doom hud bar) (On the left is health, in the center is the face, on the right is armor)

Can probably do, yeah.


> Ability to change the size of health/armor icons

I don't really understand this one either. You can change the size of the element, the icons are made to fill the existing space. You want to be able to make them smaller?

> Ability to display health/armor and ammo not only horizontally, but also vertically

Doable for ammo. For health and armor, not sure, since they're a single element and I don't really see how this would look.

> Possibility of adding several columns for all ammo display (In case there are many types of ammo, or for the compactness of the HUD)

That's a valid idea, will add.

> Add the ability to display only current ammo, or current ammo with maximum ammo. (Screen above)

Will add.

> Option to enable vertical inventory bar

I'll check how feasible it is at this stage. Might add.

> Ability to load minimap colors from the fullscreen map

Will have to check if it's possible; might not be. If it is, will add.

**Comment by SkullGamer205** [comment]
> > Increase the shift limit of the horizontal and vertical shift of an element (Up to 4096) (Or add a shift by percentage)
> 
> I'm not sure I see what the point of this is. The distance is designed to cover a specific region around the currently used anchor point. If you want to move further, it would make more sense to just change the anchor point first (move to a different corner/edge).

Yes, The distance is designed to cover a specific region around the currently used anchor point, and I can move it to the other edge, but! When the interface size is < 3, it becomes more difficult to organize the interface.
Yes, with the appearance of the "scale" option you can cheat a little, but it sounds like a crutch. I think I went overboard with 4096, 512 will be enough (especially if the HUD size = 1.

> > Add the ability to change the font of each element (not only from those existing in Doom, but from those loaded by separate WADs (By entering the font name))
> 
> Currently being worked on. Not for each element separately, though, that'd be too much work. There are 3 fonts used in the HUD, and each of them will be modifiable separately, with support for custom font names. Even that was kind of a nightmare to implement, so I'm not sure I will expand this further.

This is also enough. Even better, I think.

> > Ability to attach a character's face to the left of the character status information
> > Ability to attach the character’s face in the middle (a.k.a. classic doom hud bar) (On the left is health, in the center is the face, on the right is armor)
> 
> Can probably do, yeah.
> 
> 
> > Ability to change the size of health/armor icons
> 
> I don't really understand this one either. You can change the size of the element, the icons are made to fill the existing space. You want to be able to make them smaller?

Make them not only smaller, but and bigger.

> > Ability to display health/armor and ammo not only horizontally, but also vertically
> 
> Doable for ammo. For health and armor, not sure, since they're a single element and I don't really see how this would look.

Good.

**Comment by jekyllgrim** [comment]
> Make them not only smaller, but and bigger.

How? The icons fill all the available space. If they were made bigger, they would either clip into other bits of the health/armor block, or the whole block itself would have to be increased in size—and the latter can already be done by scaling it individually.

**Comment by SkullGamer205** [comment]
> and the latter can already be done by scaling it individually.
Oh. Really. This will be better.

**Comment by jekyllgrim** [comment]
Custom fonts and All Ammo block formatting added in e94525b7d922877cb831dfbb5f1350cbfaaa484d

**Comment by SkullGamer205** [comment]
> Custom fonts and All Ammo block formatting added in [e94525b](https://github.com/jekyllgrim/UniversalFlexibleHUD/commit/e94525b7d922877cb831dfbb5f1350cbfaaa484d)
Yep. I see. But you forgot a Vertical shift in General Ammos.
![изображение](https://github.com/jekyllgrim/UniversalFlexibleHUD/assets/61167080/46ba54be-cb7b-4949-8bfd-58cab108bf78)

**Comment by jekyllgrim** [comment]
Whoops, fixed.

**Comment by jekyllgrim** [comment]
> Ability to load minimap colors from the fullscreen map

Added in cbb4d6f6303c24d8bdd16d4cdbdb5745dfbd3d1b

**Comment by SkullGamer205** [comment]
> >  Increase the shift limit of the horizontal and vertical shift of an element (Up to 4096) (Or add a shift by percentage)
> 
> 530e8fe5a08274681a4d9dc3c655a2fdd03a2ef7 adds a new option "Use clean offset scaling" (under Appearance options). If this is enabled, the offset values will be scaled independently from the scale of the element, producing the same offsets regardless of resolution. This should solve this issue.

I'll check it tomorrow.

**Comment by jekyllgrim** [comment]
> Ability to attach a character's face to the left of the character status information
Added in d7fe72ce9327c9eda0c322c6717db23291c68ad6

> Ability to attach the character’s face in the middle (a.k.a. classic doom hud bar) (On the left is health, in the center is the face, on the right is armor)
Can't be done with the current structure. Might revisit this later, but for now consider it on hold.

**Comment by SkullGamer205** [comment]
> > Ability to attach a character's face to the left of the character status information
> 
> Added in d7fe72ce9327c9eda0c322c6717db23291c68ad6
> 
> > Ability to attach the character’s face in the middle (a.k.a. classic doom hud bar) (On the left is health, in the center is the face, on the right is armor)
> 
> Can't be done with the current structure. Might revisit this later, but for now consider it on hold.
> 

Good!

**Comment by SkullGamer205** [comment]
> > >  Increase the shift limit of the horizontal and vertical shift of an element (Up to 4096) (Or add a shift by percentage)
> > 
> > 530e8fe5a08274681a4d9dc3c655a2fdd03a2ef7 adds a new option "Use clean offset scaling" (under Appearance options). If this is enabled, the offset values will be scaled independently from the scale of the element, producing the same offsets regardless of resolution. This should solve this issue.
> 
> I'll check it tomorrow.

That works!

**Comment by jekyllgrim** [comment]
Vertical inventory bar added in 40cd0cf669c52b178af53f4b050ba619127557cd

**Comment by SkullGamer205** [comment]
> Vertical inventory bar added in 40cd0cf669c52b178af53f4b050ba619127557cd
> 
> (The current dev build has partially broken presets, I recommend waiting for a stable release)

Okay! Good!

**Comment by aeonlight** [comment]
seeing as this already exists and is still marked open, I'd like to add my own request to it.
being able to manually control the number of columns on the key display would be nice.
